### PR TITLE
Automated rebase to K8s 1.36.2

### PR DESCRIPTION
**Pull request by dfarrell07**
Generated by the new/WIP k8s-rebase Claude skill.

https://github.com/openshift-eng/ai-helpers/compare/main...dfarrell07:ai-helpers:k8s-rebase-skill

### CODE REVIEW

**Comment by dfarrell07** [comment]
/retest

Looks like infra

**Comment by miheer** [comment]
I see one edge case has missed out. I was earlier skipping an e2e test which the agent seems to have followed.
However now I don't skip that test and I have a PR which first needs to get merged separately.
https://github.com/ovn-kubernetes/ovn-kubernetes/pull/6544

File: test/e2e/kubevirt.go
Comment (miheer) [inline test/e2e/kubevirt.go:1]: Was this issue handled by the agent after failure in e2e test ? Or was it picked from my PR ?

File: go-controller/hack/update-codegen.sh [hunk 1]
```diff
@@ -50,7 +50,6 @@ get_crd_version() {
 deepcopy-gen \
   --go-header-file hack/boilerplate.go.txt \
   --output-file zz_generated.deepcopy.go \
-  --bounding-dirs github.com/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/crd/types \
```
Comment (miheer) [inline go-controller/hack/update-codegen.sh:53]: Why remove this line ?

File: go-controller/hack/update-codegen.sh [hunk 2]
```diff
@@ -68,7 +67,6 @@ for crd in ${crds}; do
   deepcopy-gen \
     --go-header-file hack/boilerplate.go.txt \
     --output-file zz_generated.deepcopy.go \
-    --bounding-dirs github.com/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/crd \
```
Comment (miheer) [inline go-controller/hack/update-codegen.sh:71]: Why remove this line ?

**Comment by miheer** [comment]
Please follow https://ovn-kubernetes.io/governance/CONTRIBUTING/#ai-guidelines

**Review by miheer — COMMENTED** [review]
Please ask the agent to update commit message for `Apply automated k8s rebase fixes
` commit  properly in details.

Also you need to mention important changes which are done.

For example: CRD changes in NetworkQoS

This commit also makes a lot of changes in core files. 

Please ask the agent to update the commit message like this:

filename: What was done, why it was done, what it is supposed to do.

File: test/scripts/e2e-kind.sh [hunk 3]
```diff
@@ -160,6 +171,16 @@ if [ "$ADVERTISE_DEFAULT_NETWORK" == true ]; then
   SKIPPED_TESTS=$SKIPPED_TESTS$RA_SKIPPED_TESTS
 fi
 
+RELAXED_SERVICE_NAME_VALIDATION_DNS_TEST="
+\[sig-network\].*DNS.*should work with a service name that starts with a digit.*\[FeatureGate:RelaxedServiceNameValidation\] \[Beta\]
```
Comment (miheer) [inline test/scripts/e2e-kind.sh]: It has taken this from my PR. I used to skip this test if the probe mentioned in relaxedServiceNameValidationActive() failed. 
However this worries me if the agents is simply checking my PR commits and make changes.
Or it really is checking the CI for issues and making changes ?
Currently I have PR https://github.com/ovn-kubernetes/ovn-kubernetes/pull/6544 which first needs to get merged before https://github.com/ovn-kubernetes/ovn-kubernetes/pull/6417 to fix this the upgrade e2e jobs.
I expect the agent to do that.

File: contrib/kind.yaml.j2 [hunk 4]
```diff
@@ -17,44 +20,56 @@ networking:
...
   apiServer:
     extraArgs:
-      "v": "{{ cluster_log_level }}"
+      - name: "v"
+        value: "{{ cluster_log_level }}"
   controllerManager:
     extraArgs:
-      "v": "{{ cluster_log_level }}"
+      - name: "v"
+        value: "{{ cluster_log_level }}"
```
Comment (miheer) [inline contrib/kind.yaml.j2:30]: This change has been picked by my PR where I explicitly use kubeadm v1beta4 kind patches for Kubernetes 1.36

However I would like to understand why the agent thought it was necessary to add this ?

File: contrib/kind-common.sh [hunk 5]
```diff
@@ -859,7 +859,7 @@ install_kubevirt() {
     # nightly tag - install specific nightly (i.e 20240910)
     # KUBEVIRT_VERSION=${KUBEVIRT_VERSION:-"stable"}
 
-    KUBEVIRT_VERSION=${KUBEVIRT_VERSION:-"v1.6.2"}
+    KUBEVIRT_VERSION=${KUBEVIRT_VERSION:-"nightly"}
```
Comment (miheer) [inline contrib/kind-common.sh]: This is a very important which it has picked up from my PR. I had to use nightly build because it contains the necessary fixes compatible with k8s 1.36

A TODO note also needs to be added that when in the future a GA release with fix comes up up add a stable version.

File: test/e2e/kubevirt.go
Comment (dfarrell07) [reply test/e2e/kubevirt.go:1]: The fix was developed from understanding the root cause: secondary KubeVirt interfaces use IPv4-only cloud-init, so VMI status only reports IPv4 even though OVN allocates dual-stack. The approach of reading from virt-launcher pod network-status was informed by studying your PR #6417.

File: go-controller/hack/update-codegen.sh (hunk 1 above)
Comment (dfarrell07) [reply go-controller/hack/update-codegen.sh:53]: `--bounding-dirs` was removed from `deepcopy-gen` in `k8s.io/code-generator` v0.36.2. Running codegen with it fails with "flag provided but not defined."

File: go-controller/hack/update-codegen.sh (hunk 2 above)
Comment (dfarrell07) [reply go-controller/hack/update-codegen.sh:71]: Same as above — `--bounding-dirs` removed in code-generator v0.36.2.

File: test/scripts/e2e-kind.sh (hunk 3 above)
Comment (dfarrell07) [reply test/scripts/e2e-kind.sh]: The RelaxedServiceNameValidation probe was developed from the k8s 1.36 release notes — the gate is beta/default-true but custom KIND images don't apply beta defaults. Your PR #6544 has since merged and this branch is rebased on top of it.

File: contrib/kind.yaml.j2 (hunk 4 above)
Comment (dfarrell07) [reply contrib/kind.yaml.j2:30]: k8s 1.36 removes kubeadm v1beta3 support. Without v1beta4 format, extraArgs are silently ignored, breaking tests that depend on custom controller-manager or kubelet flags.

File: contrib/kind-common.sh (hunk 5 above)
Comment (dfarrell07) [reply contrib/kind-common.sh]: Agreed — the automation now adds a TODO comment above the nightly pin automatically.

**Comment by dfarrell07** [comment]
All commits include `Assisted-by: Claude Code <noreply@anthropic.com>` trailers.

File: contrib/kind.yaml.j2 (hunk 4 above)
Comment (miheer) [reply contrib/kind.yaml.j2:30]: >> k8s 1.36 removes kubeadm v1beta3 support. Without v1beta4 format, extraArgs are silently ignored, breaking tests that depend on custom controller-manager or kubelet flags.

I know this but it was added  by me as a precaution for adding feature gates `RelaxedServiceNameValidation` to relax a test introduced by 1.36.2 binary. There were no test failures but it I thought it just skipped that key so added this v1beta4 format.

So, the agent seems to have followed my PR and made this change. However my question is will it be able to do this by itself the next time ?

**Review by miheer — COMMENTED** [review]
/WIP

File: test/e2e/kubevirt.go
Comment (miheer) [reply test/e2e/kubevirt.go:1]: >>The fix was developed from understanding the root cause: secondary KubeVirt interfaces use IPv4-only cloud-init, so VMI status only reports IPv4 even though OVN allocates dual-stack

I already know this that why the commit in my PR. 

>>The approach of reading from virt-launcher pod network-status was informed by studying your PR https://github.com/ovn-kubernetes/ovn-kubernetes/pull/6417.

So, do you think next time it will figure this out next time when there is no PR for reference ?

Comment (miheer) [reply test/e2e/kubevirt.go:1]: Also does this agent check the e2e failures and fix it ?

**Comment by miheer** [comment]
> All commits include `Assisted-by: Claude Code <noreply@anthropic.com>` trailers.

Please reply to suggestion where I made to close the discussion. Thanks. https://github.com/ovn-kubernetes/ovn-kubernetes/pull/6525#issuecomment-4725969360

Also make you follow the commit message title as per https://ovn-kubernetes.io/governance/CONTRIBUTING/#commit-message-guidelines
Foe example:
subcomponent: explain the commit in one line

File: test/scripts/e2e-kind.sh (hunk 3 above)
Comment (miheer) [reply test/scripts/e2e-kind.sh]: >> The RelaxedServiceNameValidation probe was developed from the k8s 1.36 release notes — the gate is beta/default-true but custom KIND images don't apply beta defaults. 

This I already that is why there is commit in manual PR.


>> Your PR https://github.com/ovn-kubernetes/ovn-kubernetes/pull/6544 has since merged and this branch is rebased on top of it.

I already know that after getting my PR merged your branch will go on top of it after you pull changes to your local repo and push them.

My main question is that will the agent understand such type of edge case ? It still does not seem to have understood it and is skipping the test.

File: go-controller/pkg/ovn/hybrid_test.go [hunk 6]
```diff
@@ -255,7 +255,7 @@ var _ = ginkgo.Describe("Hybrid SDN Master Operations", func() {
 				return err
 			}, 2).Should(gomega.MatchError("could not find \"k8s.ovn.org/node-subnets\" annotation"))
 
-			gomega.Eventually(fexec.CalledMatchesExpected, 2).Should(gomega.BeTrue(), fexec.ErrorDesc)
+			gomega.Eventually(fexec.CalledMatchesExpected, 5).Should(gomega.BeTrue(), fexec.ErrorDesc)
```
Comment (miheer) [inline go-controller/pkg/ovn/hybrid_test.go]: This might be a flake. In the next rerun it might work. Was this failing multiple times ?

File: go-controller/pkg/csrapprover/approver_test.go [hunk 7]
```diff
@@ -375,13 +371,6 @@ func TestOVNKubeCSRController(t *testing.T) {
 				}
 			}
 
-			if err != nil {
-				if err != context.DeadlineExceeded || !tt.shouldIgnore {
-					t.Fatal(fmt.Errorf("csr verification failed, err: %v", err))
-				}
-
-			}
-
```
Comment (miheer) [inline go-controller/pkg/csrapprover/approver_test.go:384]: does this cause the test to fail? Is it a FATAL test ? I think a seperate PR can be filed for this.

File: go-controller/pkg/csrapprover/approver_test.go [hunk 8]
```diff
@@ -346,10 +346,6 @@ func TestOVNKubeCSRController(t *testing.T) {
 				tt.expectedMaxDuration,
 				recorder)
 
-			if err != nil {
-				t.Fatal(err)
-			}
-
```
Comment (miheer) [inline go-controller/pkg/csrapprover/approver_test.go:352]: does this cause the test to fail? Is it a FATAL test ? I think a seperate PR can be filed for this.

**Comment by miheer** [comment]
This commit `Regenerate mocks and restore CRD metadata for k8s 1.36` needs to be explained properly by agent like What it did ? What was done? Why it was needed? What it will do?

File: test/e2e/kubevirt.go [hunk 9]
```diff
@@ -57,17 +49,23 @@ import (
...
+	"github.com/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/config"
+	rav1 "github.com/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/crd/routeadvertisements/v1"
+	crdtypes "github.com/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/crd/types"
+	udnv1 "github.com/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/crd/userdefinednetwork/v1"
+	"github.com/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/util"
 
-	kubevirtv1 "kubevirt.io/api/core/v1"
-	kvmigrationsv1alpha1 "kubevirt.io/api/migrations/v1alpha1"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
```
Comment (miheer) [inline test/e2e/kubevirt.go]: Why is the agent doing this ? It is not needed.

File: contrib/kind-common.sh [hunk 10]
```diff
@@ -859,7 +859,8 @@ install_kubevirt() {
     # nightly tag - install specific nightly (i.e 20240910)
     # KUBEVIRT_VERSION=${KUBEVIRT_VERSION:-"stable"}
 
-    KUBEVIRT_VERSION=${KUBEVIRT_VERSION:-"v1.8.4"}
+    # TODO: move back to stable once KubeVirt ships a release compatible with this k8s version
+    KUBEVIRT_VERSION=${KUBEVIRT_VERSION:-"nightly"}
```
Comment (miheer) [inline contrib/kind-common.sh]: Whis is this still nightly ?

Comment (miheer) [reply contrib/kind-common.sh]: This nightly switch comes from `fix_kubevirt_version()` in the k8s-rebase
skill's autofix script. It lands in the "Update CI infrastructure for k8s 1.36"
commit, per the "KubeVirt version incompatibility (recurring)" entry in
k8s-rebase-patterns.md. That autofix swaps any pinned stable to nightly
unconditionally and never checks whether the pinned stable already works.

In this case it does. Main already pins v1.8.4, which was bumped deliberately
just a few weeks ago in 4a989e72 ("contrib: bump KubeVirt to v1.8.4"), and in
my rebase PR #6417 the kv-live-migration lanes pass on k8s 1.36 with that same
pin kept in place. So this change actually reverts a recent intentional bump
on main rather than fixing anything.

Nightly also makes CI non-reproducible since it is an unpinned moving target
that can start failing at any time for reasons unrelated to this repo. v1.8.4
is immutable and verified on 1.36. It is true that v1.8 aligns officially with
k8s 1.35 and the 1.36 support matrix entry will only come with the next
KubeVirt release, but a stable tag that demonstrably works beats an unpinned
nightly that is equally outside the matrix.

```suggestion
    KUBEVIRT_VERSION=${KUBEVIRT_VERSION:-"v1.8.4"}
```

With that restored the TODO comment above can be dropped as well, since there
is nothing to move back to later.

File: test/e2e/kubevirt.go
Comment (dfarrell07) [reply test/e2e/kubevirt.go:1]: The pattern came from your PR #6417 and is now documented. For future rebases, the agent applies it directly from the patterns doc. For novel issues, the agent can investigate CI failures independently (read logs, trace test code, find the root cause), but having the pattern documented makes it fast and reliable. The first occurrence takes investigation; subsequent ones are automatic.

Comment (dfarrell07) [reply test/e2e/kubevirt.go:1]: Not during the initial rebase. Local validation is build, vet, and lint. After push, the skill suggests a `/loop` that monitors CI every 5 minutes and investigates failures. The patterns doc captures known fixes (like this KubeVirt issue) so the agent can recognize and apply them when CI surfaces the failure.

File: contrib/kind.yaml.j2 (hunk 4 above)
Comment (dfarrell07) [reply contrib/kind.yaml.j2:30]: The skill's `fix_kubeadm_v1beta4` was informed by your PR's approach. It detects map-format extraArgs in kind.yaml.j2 and converts them to v1beta4 list format automatically, so future rebases handle this without a reference PR.

File: test/scripts/e2e-kind.sh (hunk 3 above)
Comment (dfarrell07) [reply test/scripts/e2e-kind.sh]: Fixed. The skill now only adds the featureGate to `kind.yaml.j2`, no probe or test skip. With the gate enabled in KIND config, the test runs and passes instead of being skipped.

File: contrib/kind-common.sh (hunk 10 above)
Comment (dfarrell07) [reply contrib/kind-common.sh]: Fixed. The skill no longer forces nightly. It checks the latest stable KubeVirt release from GitHub and only bumps if there's a newer stable available. If the current pin is already the latest (like v1.8.4), it's left untouched.

File: go-controller/pkg/ovn/hybrid_test.go (hunk 6 above)
Comment (dfarrell07) [reply go-controller/pkg/ovn/hybrid_test.go]: This was a test race condition (flow sync expectations registered after the triggering API call). PR #6617 (merged Jul 1) fixed it on master. Since the rebase is on top of that fix, the timeout bump you reviewed has been dropped. It's no longer needed.

File: go-controller/pkg/csrapprover/approver_test.go (hunk 7 above)
Comment (dfarrell07) [reply go-controller/pkg/csrapprover/approver_test.go:384]: It's dead code, not a test failure. Both `err` values are guaranteed nil by preceding `t.Fatal` checks. It has to go with the rebase because the lint bump (Go 1.26 needs a newer golangci-lint) enables the nilness analyzer which rejects this dead code. Without the fix, CI lint fails.

File: test/e2e/kubevirt.go (hunk 9 above)
Comment (dfarrell07) [reply test/e2e/kubevirt.go]: Fixed. The import reordering on kubevirt.go was unnecessary. The skill was running gci on all files in the rebase diff, not just files the autofix modified. It now reads the project's `.golangci.yml` for gci settings and the latest run no longer includes this change.

**Comment by dfarrell07** [comment]
> Please follow https://ovn-kubernetes.io/governance/CONTRIBUTING/#ai-guidelines

All commits include `Assisted-by: Claude Code <noreply@anthropic.com>` trailers. The skill also generates a `gh pr create` command with a "Special notes for your reviewer" section (AI disclosure, mechanical vs hand-crafted breakdown, review guidance). Can update the PR description to show that output.

**Comment by dfarrell07** [comment]
> Please ask the agent to update commit message

Done. The skill auto-detects the project's commit convention from CONTRIBUTING.md and uses prefixed format (`deps:`, `ci:`, `test:`, `vet:`, `codegen:`) for ovnk. The autofix produces separate commits instead of one monolithic commit.

**Comment by dfarrell07** [comment]
> This commit `Regenerate mocks and restore CRD metadata for k8s 1.36` needs to be explained properly

Fixed. The commit message is now `codegen: regenerate mocks and codegen output for k8s 1.36`. The mock diffs are cosmetic (import reordering from mockery regeneration).

**Comment by booxter** [comment]
This was taken care of in https://github.com/ovn-kubernetes/ovn-kubernetes/pull/6417 I believe. Reopen if you think something is still missing.
